logos-derive/src/parser/subpattern.rs

```diff
@@ -0,0 +1,118 @@
+use proc_macro2::TokenStream;
+use syn::{Ident, LitByteStr, LitStr};
+
+use crate::error::Errors;
+use crate::mir::Mir;
+use crate::parser::definition::{bytes_to_regex_string, Literal};
+
+#[derive(Default)]
+pub struct Subpatterns {
+    map: Vec<(Ident, Literal)>,
+}
+
+impl Subpatterns {
+    pub fn add(&mut self, param: Ident, pattern: TokenStream, errors: &mut Errors) {
+        let lit = match syn::parse2::<Literal>(pattern) {
+            Ok(lit) => lit,
+            Err(e) => {
+                errors.err(e.to_string(), e.span());
+                return;
+            }
+        };
+
+        if let Some((name, _)) = self.map.iter().find(|(name, _)| *name == param) {
+            errors
+                .err(format!("{} can only be assigned once", param), param.span())
+                .err("Previously assigned here", name.span());
+            return;
+        }
+
+        // Validate the literal as proper regex. If it's not, error and manufacture a substitute.
+        let lit = match &lit {
+            Literal::Utf8(s) => match Mir::utf8(&s.value()) {
+                Ok(_) => lit,
+                Err(err) => {
+                    errors.err(err, lit.span());
+                    Literal::Utf8(LitStr::new(&param.to_string(), lit.span()))
+                }
+            },
+            Literal::Bytes(b) => {
+                let source = bytes_to_regex_string(b.value());
+                match Mir::binary(&source) {
+                    Ok(_) => lit,
+                    Err(err) => {
+                        errors.err(err, lit.span());
+                        Literal::Bytes(LitByteStr::new(param.to_string().as_bytes(), lit.span()))
+                    }
+                }
+            }
+        };
+
+        self.map.push((param, lit));
+    }
+
+    pub fn fix(&self, lit: &Literal, errors: &mut Errors) -> String {
+        let mut i = 0;
+        let mut pattern = match lit {
+            Literal::Utf8(s) => s.value(),
+            Literal::Bytes(b) => bytes_to_regex_string(b.value()),
+        };
+
+        while i < pattern.len() {
+            if let Some(f) = pattern[i..].find("(?&") {
+                i += f;
+                pattern.replace_range(i..i + 3, "(?:");
+                i += 3;
+            } else {
+                break; // done!
+            }
```

Comment:
```suggestion
        while let Some(f) = pattern[i..].find("(?&") {
            i += f;
            pattern.replace_range(i..i + 3, "(?:");
            i += 3;
```
+ I don't think we need to worry about `i` going out of bounds here